feat(extensions): handle build step for expressjs framework (#1315)
canonical/rockcraft: docs/reference/extensions/express-framework.rst

Reviewer: Slightly out of scope for this PR, but since we're updating this page anyway, could we include the link for the Ubuntu package repository on this line?
```suggestion
If you don't customize the version of node, it will be installed from the
`Ubuntu package repository <https://packages.ubuntu.com/noble/nodejs>`_.
```

Express framework
=================

The Express extension streamlines the process of building Express

> application rocks.

Reviewer: I don't think we should remove this line. Please keep it before the tabbed environment

It facilitates the installation of Express application dependencies, including

> Node.js and npm, inside the rock. Additionally, it transfers your project files

Reviewer: It looks like `It facilitates the installation of Express application dependencies, including Node.js and npm, inside the rock.` is the same for all Ubuntu versions. Please pull this sentence out of the tabbed environment and place it alongside the first sentence :)

          Node.js and npm, inside the rock. Additionally, it transfers your project files
          to ``/app`` within the rock.

          By default, the system foundation, or base, is set as ``bare`` to generate a
          lightweight image.
          The Express extension is compatible with the ``bare`` and ``ubuntu@24.04``
          bases.

    .. tab-item:: Ubuntu 26.04 and higher
        :sync: base-26-plus
          The Express extension streamlines the process of building Express
          application rocks.

          It facilitates the installation of Express application dependencies, including
          Node.js and npm, inside the rock. The extension discovers the location of the
          ``package.json``, but the extension can only package a single application.

          By default, the system foundation, or base, is set as ``bare`` to generate a
          lightweight image.
          The Express extension is compatible with the ``bare`` and ``ubuntu@24.04``
          bases.

.. _reference-express-framework-project-requirements:

Project requirements
--------------------

.. tab-set::
    .. tab-item:: Ubuntu 22.04 and 24.04
        :sync: base-22-24

>           There are 3 requirements to be able to use the ``expressjs-framework``

Reviewer: ```suggestion
          There are three requirements to be able to use the ``expressjs-framework``
```

=================

The Express extension streamlines the process of building Express
application rocks. It facilitates the installation of Express application dependencies, including
Node.js and npm, inside the rock.

.. tab-set::
    .. tab-item:: Ubuntu 22.04 and 24.04
        :sync: base-22-24

        Additionally, it transfers your project files
        to ``/app`` within the rock.

    .. tab-item:: Ubuntu 26.04 and higher
        :sync: base-26-plus

        The extension discovers the location of the
        ``package.json``, but the extension can only package a single application.

By default, the system foundation, or base, is set as ``bare`` to generate a
lightweight image.

> bases.

Reviewer 2: ```suggestion
```
Probably an artifact from a previous edit

The Express extension is compatible with the ``bare``, ``ubuntu@24.04`` and ``ubuntu@26.04``
bases.

.. _reference-express-framework-project-requirements:

Project requirements
--------------------

.. tab-set::
    .. tab-item:: Ubuntu 22.04 and 24.04
        :sync: base-22-24

        There are three requirements to be able to use the ``expressjs-framework``
        extension:

        1. The application should reside in the ``app`` directory.
        2. The application should have a ``package.json`` file.

>         3. The ``package.json`` file should define the ``start`` script.

Reviewer 2: ```suggestion
        - The application should reside in the ``app`` directory.
        - The application should have a ``package.json`` file.
        - The ``package.json`` file should define the ``start`` script.
```
Since these list items aren't strictly sequential, we don't need to enumerate them.

        For more information, see the `npm documentation <https://docs.npmjs.com/cli/v11/configuring-npm/package-json>`_.

    .. tab-item:: Ubuntu 26.04 and higher
        :sync: base-26-plus

        There are two requirements to be able to use the ``expressjs-framework``
        extension:

        1. The application should have a ``package.json`` file in the same
        directory as ``rockcraft.yaml`` file or 1 directory below the ``rockcraft.yaml`` file.

>         2. The ``package.json`` file should define the ``start`` script.

Reviewer 2: ```suggestion
        - The application should have a ``package.json`` file in the same
        directory as ``rockcraft.yaml`` file or 1 directory below the ``rockcraft.yaml`` file.
        - The ``package.json`` file should define the ``start`` script.
```
Same as https://github.com/canonical/rockcraft/pull/1315/changes#r3799280695

        For more information, see the `npm documentation <https://docs.npmjs.com/cli/v11/configuring-npm/package-json>`_.

>         If the application defines a ``build`` script in ``package.json`` file, it is

Reviewer 2: ```suggestion
        If the application defines a ``build`` script in its ``package.json`` file, it is
```

application rocks.

It facilitates the installation of Express application dependencies, including

> Node.js and npm, inside the rock. Extension discovers location of the

Reviewer 3: ```suggestion
Node.js and npm, inside the rock. The extension discovers the location of the
```

> ``package.json``, but can only package a single application. If application

Reviewer 3: ```suggestion
``package.json``, but can only package a single application. If the application
```

> ``package.json``, but can only package a single application. If the application

Reviewer: ```suggestion
Node.js and npm, inside the rock. The extension discovers the location of the
``package.json``, but it can only package a single application. If the application
```
or (to be more explicit, albeit repetitive):

```suggestion
Node.js and npm, inside the rock. The extension discovers the location of the
``package.json``, but the extension can only package a single application. If the application
```

> defines ``build`` script, development dependencies will be installed,

Reviewer 3: Usage of "a" or "the" depends on if there is only one specific build script or not. Which case is this?
```suggestion
defines a ``build`` script, development dependencies will be installed,
```

> ``npm run build`` called, and entries matching ``file`` array (excluding

Reviewer 3: Adding verb for consistency with the other two items.
```suggestion
``npm run build`` will be called, and entries matching ``file`` array (excluding
```

> entries from ``.npmignore``, if exists) will be packages. If ``files``

Reviewer 3: ```suggestion
entries from ``.npmignore``, if they exist) will be packages. If the ``files``
```

array is not defined and ``.npmignore`` does not exist, only files from

> the ``dist/`` directory will be packaged.

Reviewer: Nitpick: This new addition feels more like it belongs under "Project requirements" rather than the introduction to the Reference page. Why do users need to know about this information *first*, before the project requirements?

To be clear, I think this information is important to surface, but I'm not sure if we should be surfacing this information in the second paragraph of the page.

Project requirements
--------------------

There are two requirements to be able to use the ``expressjs-framework``
extension:

1. The application should have a ``package.json`` file.
2. The ``package.json`` file should define the ``start`` script.
   For more information, see the `npm documentation <https://docs.npmjs.com/cli/v11/configuring-npm/package-json>`_.

> If application defines ``build`` script in ``package.json`` file, it is

Reviewer 3: For using "a" or "the" in front of build script, see comment above.
```suggestion
If the application defines a ``build`` script in ``package.json`` file, it is
```

> recommended to have ``files`` array describing the entries to be included

Reviewer 3: ```suggestion
recommended to have a ``files`` array describing the entries to be included
```

or have an appropriate ``.npmignore`` file to exclude entries not required

> at the run time. If ``files`` array is not defined and ``.npmignore``

Reviewer 3: ```suggestion
at the run time. If the ``files`` array is not defined and ``.npmignore``
```

Reviewer: Adding onto this suggestion:
```suggestion
at runtime. If the ``files`` array is not defined and ``.npmignore``
```
